attribution/api_attribution.py

```diff
@@ -108,142 +310,62 @@ def make_output_location_invariant(self, original_output, perturbed_output):
-            ):
-                batch = [
-                    tasks[i] for i in range(idx, min(idx + self.request_chunksize, len(tasks)))
-                ]
-                outputs.extend(await asyncio.gather(*batch))
-                await asyncio.sleep(0.1)
         else:
-            outputs = await asyncio.gather(*tasks)
-
-        for perturbation, perturbed_output in zip(perturbations, outputs):
-            if ignore_output_token_location:
-                perturbed_output = self.make_output_location_invariant(
-                    original_output, perturbed_output
-                )
+            raise ValueError(f"Unknown attribution strategy: {attribution_strategy}")
 
-            for attribution_strategy in attribution_strategies:
-                if attribution_strategy == "cosine":
-                    sentence_attr, attributed_tokens, token_attributions = (
-                        cosine_similarity_attribution(
-                            original_output.message.content,
-                            perturbed_output.message.content,
-                            self.token_embeddings,
-                            self.tokenizer,
-                        )
-                    )
-                elif attribution_strategy == "prob_diff":
-                    sentence_attr, attributed_tokens, token_attributions = token_prob_attribution(
-                        original_output.logprobs, perturbed_output.logprobs
-                    )
-                else:
-                    raise ValueError(f"Unknown attribution strategy: {attribution_strategy}")
-
-                if logger:
-                    for i, unit_token in enumerate(perturbation["unit_tokens"]):
-                        logger.log_input_token_attribution(
-                            attribution_strategy,
-                            unit_offset + i,
-                            unit_token,
-                            float(sentence_attr),
-                        )
-                        for j, attr_score in enumerate(token_attributions):
-                            logger.log_token_attribution_matrix(
-                                attribution_strategy,
-                                unit_offset + i,
-                                j,
-                                attributed_tokens[j],
-                                attr_score.squeeze(),
-                                perturbation["input"],
-                                perturbed_output.message.content,
-                            )
-            unit_offset += len(perturbation["unit_tokens"])
+        scores = {
+            "sentence_attribution": np.mean(list(token_attributions.values())),
+            "token_attribution": token_attributions,
+        }
+        norm_scores = {
+            "sentence_attribution": np.mean(list(token_attributions.values())) / chunksize,
+            "token_attribution": {k: v / chunksize for k, v in token_attributions.items()},
```

Comment:
Does this take into account edge chunks (e.g. when/if chunk is less than chunk_size)?

Reply:
Yes! Chunksize comes from `sum(mask)` on the outside in `hierarchical_perturbation`, and for `compute_attributions` the chunksize is always 1.